### PR TITLE
8265940: Enable C2's optimization for Math.pow(x, 0.5) on all platforms

### PR DESCRIPTION
**Pull request by DamonFool**
Hi all,

I'd like to enable C2's optimization for Math.pow(x, 0.5) on all platforms.

This is fine because:
 1) For x86, it will call the assembly stubs.
    This opt has already been implemented for x86_64 [1]. 
    And the x86_32 version is added in this patch.
 2) For non-x86, it will call the shared runtime c code [2], which does the same opt too.

Testing:
  - tier1~3 on Linux/{x86_32, x86_64}, no regression

Thanks,
Best regards,
Jie

[1] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/x86/macroAssembler_x86_pow.cpp#L838
[2] https://github.com/openjdk/jdk/blob/master/src/hotspot/share/runtime/sharedRuntimeTrans.cpp#L497

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8265940](https://bugs.openjdk.java.net/browse/JDK-8265940): Enable C2's optimization for Math.pow(x, 0.5) on all platforms

### Reviewers
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**) ⚠️ Review applies to bb88f24f1c07632724d1903747b80eeb2c5be2b6
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3755/head:pull/3755` \
`$ git checkout pull/3755`

Update a local copy of the PR: \
`$ git checkout pull/3755` \
`$ git pull https://git.openjdk.java.net/jdk pull/3755/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3755`

View PR using the GUI difftool: \
`$ git pr show -t 3755`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3755.diff">https://git.openjdk.java.net/jdk/pull/3755.diff</a>

</details>

### CODE REVIEW

**Review by neliasso — APPROVED** [review]
Looks good.

**Comment by DamonFool** [comment]
> Looks good.

Thanks @neliasso .

File: src/hotspot/cpu/x86/macroAssembler_x86_pow.cpp [hunk 1]
```diff
@@ -2562,6 +2566,16 @@ void MacroAssembler::fast_pow(XMMRegister xmm0, XMMRegister xmm1, XMMRegister xm
   jmp(L_2TAG_PACKET_21_0_2);
 
   bind(L_NOT_DOUBLE2);
+  // Special case: pow(x, 0.5) => sqrt(x)
+  ucomisd(xmm1, ExternalAddress(DOUBLE0DOT5)); // For pow(x, y), check whether y == 0.5
+  jccb(Assembler::notEqual, L_NOT_DOUBLE0DOT5);
+  jccb(Assembler::parity, L_NOT_DOUBLE0DOT5);
+  ucomisd(xmm0, ExternalAddress(DOUBLE0));
+  jccb(Assembler::belowEqual, L_NOT_DOUBLE0DOT5); // pow(x, 0.5) => sqrt(x) only for x > 0.0
```
Comment (vnkozlov) [inline src/hotspot/cpu/x86/macroAssembler_x86_pow.cpp:2574]: Add comment about why 0. case is skipped: -0.0 case.

File: src/hotspot/share/opto/library_call.cpp [hunk 2]
```diff
@@ -1643,9 +1643,7 @@ bool LibraryCallKit::inline_math_pow() {
       Node* base = round_double_node(argument(0));
       set_result(_gvn.transform(new MulDNode(base, base)));
       return true;
-    }
-#if defined(X86) && defined(_LP64)
-    else if (d->getd() == 0.5 && Matcher::match_rule_supported(Op_SqrtD)) {
+    } else if (d->getd() == 0.5 && Matcher::match_rule_supported(Op_SqrtD)) {
```
Comment (vnkozlov) [inline src/hotspot/share/opto/library_call.cpp:1646]: Add the same comment at line #1654 explaining why 0. is excluded.

**Review by vnkozlov — CHANGES_REQUESTED** [review]
The fix looks good.

File: src/hotspot/cpu/x86/macroAssembler_x86_pow.cpp (hunk 1 above)
Comment (DamonFool) [reply src/hotspot/cpu/x86/macroAssembler_x86_pow.cpp:2577]: > Add comment about why 0. case is skipped: -0.0 case.

Thanks @vnkozlov for your review.

Comments have been added.
Thanks.

**Review by vnkozlov — APPROVED** [review]
Good.

**Pull request closed without merging**